### ISSUE TITLE
Seldon controller does not create Ambassador REST config if a non PU container exists

### ISSUE THREAD

**Issue by kparaju** [SeldonIO/seldon-operator#66]
If I add container in my pod, the seldon controller does not create the ambassador REST config.  I'm talking about the Mapping object that ends with a `-main` that has the REST and GRPC mapping. My custom container does listen on a TCP Port (not http or grpc) and we use it for streaming logs.

This object: https://github.com/SeldonIO/seldon-operator/blob/e52f9ba48f6da426f3b04e67cf61abd355c600f7/pkg/controller/seldondeployment/ambassador.go#L77

This has to be a fresh new SeldonDeployment. If you are just updating an existing one and add the second container, you will run into a weird issue where there are two deployments for that Mapping object and the traffic splits between the two deployments.

Here are the logs of the controller (truncated to remove the huge lines -- I can provide them if needed):

```
{"level":"info","ts":1569889753.1269317,"logger":"seldon-controller","msg":"pSvcName","val":"sort-server-sort-main"}
{"level":"info","ts":1569889753.1271641,"logger":"seldon-controller","msg":"Not creating container service for scribe"}
{"level":"info","ts":1569889753.1274755,"logger":"seldon-controller","msg":"Creating default Ambassador config"}
{"level":"info","ts":1569889753.1278055,"logger":"seldon-controller","msg":"Creating Deployment","namespace":"dspe-seldon","name":"sort-main-00ed72d"}
{"lvel":"info","ts":1569889760.0835779,"logger":"seldon-controller","msg":"Creating Service","namespace":"dspe-seldon","name":"seldon-447e09a2c0b7151d4e8d26ba14f046eb"}
{"level":"info","ts":1569889760.184504,"logger":"seldon-controller","msg":"Skipping Ambassador Svc"}
{"level":"info","ts":1569889760.214855,"logger":"seldon-controller","msg":"pSvcName","val":"sort-server-sort-main"}
{"level":"info","ts":1569889760.2150455,"logger":"seldon-controller","msg":"Not creating container service for scribe"}
{"level":"info","ts":1569889760.2153084,"logger":"seldon-controller","msg":"Creating default Ambassador config"}
{"level":"info","ts":1569889760.216094,"logger":"seldon-controller","msg":"Updating Deployment","namespace":"dspe-seldon","name":"sort-main-00ed72d"}
{"level":"info","ts":1569889760.239691,"logger":"seldon-controller","msg":"Found identical Service","namespace":"dspe-seldon","name":"seldon-447e09a2c0b7151d4e8d26ba14f046eb","status":{"loadBalancer":{}}}
{"level":"info","ts":1569889760.304992,"logger":"seldon-controller","msg":"Skipping Ambassador Svc"}
{"level":"info","ts":1569889760.3273509,"logger":"seldon-controller","msg":"pSvcName","val":"sort-server-sort-main"}
{"level":"info","ts":1569889760.3275404,"logger":"seldon-controller","msg":"Not creating container service for scribe"}
{"level":"info","ts":1569889760.327765,"logger":"seldon-controller","msg":"Creating default Ambassador config"}
{"level":"info","ts":1569889760.3282974,"logger":"seldon-controller","msg":"Updating Deployment","namespace":"dspe-seldon","name":"sort-main-00ed72d"}
{"level":"info","ts":1569889760.3377779,"logger":"seldon-controller","msg":"Found identical Service","namespace":"dspe-seldon","name":"seldon-447e09a2c0b7151d4e8d26ba14f046eb","status":{"loadBalancer":{}}}
{"level":"info","ts":1569889760.3378348,"logger":"seldon-controller","msg":"Skipping Ambassador Svc"}
{"level":"info","ts":1569889760.4085755,"logger":"seldon-controller","msg":"pSvcName","val":"sort-server-sort-main"}
{"level":"info","ts":1569889760.4087677,"logger":"seldon-controller","msg":"Not creating container service for scribe"}
{"level":"info","ts":1569889760.409045,"logger":"seldon-controller","msg":"Creating default Ambassador config"}
{"level":"info","ts":1569889760.4097595,"logger":"seldon-controller","msg":"Updating Deployment","namespace":"dspe-seldon","name":"sort-main-00ed72d"}
{"level":"info","ts":1569889760.4594553,"logger":"seldon-controller","msg":"Found identical Service","namespace":"dspe-seldon","name":"seldon-447e09a2c0b7151d4e8d26ba14f046eb","status":{"loadBalancer":{}}}
{"level":"info","ts":1569889760.459514,"logger":"seldon-controller","msg":"Skipping Ambassador Svc"}
{"level":"info","ts":1569889760.7984533,"logger":"seldon-controller","msg":"pSvcName","val":"sort-server-sort-main"}
{"level":"info","ts":1569889760.7986178,"logger":"seldon-controller","msg":"Not creating container service for scribe"}
{"level":"info","ts":1569889760.798862,"logger":"seldon-controller","msg":"Creating default Ambassador config"}
{"level":"info","ts":1569889760.7995014,"logger":"seldon-controller","msg":"Updating Deployment","namespace":"dspe-seldon","name":"sort-main-00ed72d"}
{"level":"info","ts":1569889760.874102,"logger":"seldon-controller","msg":"Found identical Service","namespace":"dspe-seldon","name":"seldon-447e09a2c0b7151d4e8d26ba14f046eb","status":{"loadBalancer":{}}}
{"level":"info","ts":1569889760.8741508,"logger":"seldon-controller","msg":"Skipping Ambassador Svc"}
{"level":"info","ts":1569889812.5688767,"logger":"seldon-controller","msg":"pSvcName","val":"sort-server-sort-main"}
{"level":"info","ts":1569889812.5691185,"logger":"seldon-controller","msg":"Not creating container service for scribe"}
{"level":"info","ts":1569889812.5694637,"logger":"seldon-controller","msg":"Creating default Ambassador config"}
{"level":"info","ts":1569889812.570288,"logger":"seldon-controller","msg":"Updating Deployment","namespace":"dspe-seldon","name":"sort-main-00ed72d"}
{"level":"info","ts":1569889817.0923462,"logger":"seldon-controller","msg":"Found identical Service","namespace":"dspe-seldon","name":"seldon-447e09a2c0b7151d4e8d26ba14f046eb","status":{"loadBalancer":{}}}
{"level":"info","ts":1569889817.092424,"logger":"seldon-controller","msg":"Skipping Ambassador Svc"}
```

```
(base) ➜  ~ kubectl get all
NAME                       AGE
deploy/sort-main-00ed72d   5m

NAME                              AGE
rs/sort-main-00ed72d-6c9b4698d9   5m

NAME                                    READY     STATUS    RESTARTS   AGE
po/sort-main-00ed72d-6c9b4698d9-pqw5m   3/3       Running   0          5m

NAME                                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
svc/seldon-447e09a2c0b7151d4e8d26ba14f046eb   ClusterIP   100.77.53.191   <none>        9000/TCP   5m
```

Here's the SeldonDeployment (truncated volumemounts and other irrelevant stuff)

```
---
apiVersion: machinelearning.seldon.io/v1alpha2
kind: SeldonDeployment
metadata:
  labels:
    app: seldon
    app.kubernetes.io/instance: dspe-seldon-ndmad2
    app.kubernetes.io/managed-by: argocd
    app.kubernetes.io/part-of: dspe-seldon
  name: sort-server
  namespace: dspe-seldon
spec:
  name: sort
  predictors:
  - componentSpecs:
    - spec:
        containers:
        - image: classifier:2.0
          imagePullPolicy: Always
          name: sort-classifier
          resources:
            limits:
              cpu: 4
              memory: 8Gi
          image: logger:0.14
          imagePullPolicy: IfNotPresent
          lifecycle:
            preStop:
              exec:
                command:
                - /bin/sleep
                - "15"
          name: scribe
          ports:
          - containerPort: 1463
            name: probe-port
            protocol: TCP
          - containerPort: 1473
            protocol: TCP
    graph:
      children: []
      endpoint:
        type: REST
      name: sort-classifier
      type: MODEL
    name: main
    replicas: 1
    svcOrchSpec:
      env:
      - name: JAVA_OPTS
        value: -server -Xms512m -Xmx512m -XX:+AlwaysPreTouch -XX:+UnlockExperimentalVMOptions
          -XX:G1NewSizePercent=20 -XX:+UseG1GC -XX:MaxGCPauseMillis=10 -XX:GCLogFileSize=10485760
          -XX:NumberOfGCLogFiles=1 -XX:+PrintGC -XX:+PrintGCApplicationStoppedTime
          -XX:+PrintGCDateStamps -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintTenuringDistribution
          -XX:+UseGCLogFileRotation -Xloggc:/tmp/gc.log -XX:+UseTLAB -XX:+DisableExplicitGC
      - name: SELDON_LOG_LEVEL
        value: DEBUG
      resources:
        limits:
          cpu: 4
          memory: 4Gi
        requests:
          cpu: 500m
          memory: 512Mi
```

Let me know if I can help debug in any way!

**Comment by cliveseldon** [comment]
Thanks for the detailed report.

 *  Can you state which version of Seldon you are running?
 * The yaml above seems malformed as you don't start the second container *without* a `- ` so the two containers are actually defined as one. However, when I try this it fails validation so maybe that is just a typo in the paste above?
 * It looks to me like it may be continually trying to update the deployment. There is an issue that if K8s adds some default fields we are not expecting it could be we keep thinking the deployment has changed. This maybe actually the volumeMounts you have not shown. Can you show the full yaml? If it updates the dpeloyment it should print the two version it thinks are different. Can you check how they differ by for example using a tool like: http://jsondiff.com/ or can you paste the two in the issue
 * We have actually merged the Operator into the main repo now so it would be good to test against that if you have a chance.
* I don't think this is to do with the extra container per se as I tried that and its fine with yaml similar to above

**Comment by kparaju** [comment]
Version 0.4.0, we will see if there are any changes when we upgrade to 0.4.1

Sent the YAML via PM. I'll try modifying the SeldonDeployment (removing volumemount etc) and post updates.

**Comment by cliveseldon** [comment]
I think this fix in 0.4.1 would fix this: https://github.com/SeldonIO/seldon-operator/commit/00d9d66059287cbfb4f69632e1892a4f44e9c5ed

**Comment by kparaju** [comment]
v0.4.1 fixes this.

Thanks!
